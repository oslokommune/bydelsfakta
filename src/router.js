import Vue from 'vue';
import Router from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

const District = () => import('./views/District');
const Topic = () => import('./views/Topic');
const NotFound = () => import('./views/NotFound');

Vue.use(Router);

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: District,
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: District,
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: Topic,
    props: true,
  },
  {
    path: '*',
    name: 'NotFound',
    component: NotFound,
  },
];

const router = new Router({
  mode: 'history',
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from, next) => {
  if (to.params.topic) {
    if (!topicNames.find(name => name === to.params.topic) || disabledTopics.includes(to.params.topic)) {
      next({ name: 'NotFound', params: [to.path] });
    }
  }

  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      next();
    } else if (districts.length === 1) {
      if (allDistricts.find(district => district.uri === districts[0])) {
        next();
      } else if (allDistricts.find(district => district.key === districts[0])) {
        next();
      } else {
        next({ name: 'NotFound', params: [to.path] });
      }
    } else if (districts.length > 1) {
      const errors = [];
      districts.forEach(district => {
        if (!allDistricts.find(item => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        next({ name: 'NotFound', params: [to.path] });
      }
    }
  }

  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }

  next();
});

export default router;
