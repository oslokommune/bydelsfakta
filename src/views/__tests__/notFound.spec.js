import 'whatwg-fetch';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import VueMeta from 'vue-meta';

import NotFound from '../NotFound';
import setupI18n from '../../i18n';
import { routes } from '../../router';
import mockStore from '../../../tests/MockStore';

const i18n = setupI18n();

describe('Topic', () => {
  let wrapper = null;
  let router = null;
  let store = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    localVue.use(Vuex);
    localVue.use(VueMeta);
    router = new VueRouter({ routes });
    store = new Vuex.Store(mockStore);
    wrapper = mount(NotFound, {
      localVue,
      router,
      store,
      i18n,
      stubs: {
        'v-category': true,
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
