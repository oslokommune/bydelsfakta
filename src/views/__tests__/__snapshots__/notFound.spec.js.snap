// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic renders component correctly 1`] = `
<div
  class="notFound-container"
>
  <div
    class="notFound-container--row"
  >
    <div
      class="notFound__square--outer"
    >
      <div
        class="notFound__square--inner"
      >
        <label
          class="notFound__square--label"
        >
          404
        </label>
      </div>
    </div>
     
    <div
      class="notFound__information"
    >
      <h2
        class="notFound__information--header"
      >
        Siden ble ikke funnet.
      </h2>
       
      <p>
        Du har ved en feiltakelse blitt sendt til en side som ikke eksiterer.
      </p>
       
      <a
        class="notFound__information--frontpage-link router-link-exact-active router-link-active"
        href="#/"
      >
        
        Gå til forsiden
      
      </a>
    </div>
  </div>
   
  <div
    class="notFound-container--column"
  >
    <h2
      class="notFound__information--header"
    >
      Pssst...
    </h2>
     
    <p>
      
      Vi vil gjerne høre fra deg om hva som gikk galt slik at vi kan forbedre tjenesten. Du kan sende en e-post til 
      
      <a
        href="mailto:bydelsfakta@byr.oslo.kommune.no?Subject=404%20Bydelsfakta"
        target="_top"
      >
        bydelsfakta@byr.oslo.kommune.no
      </a>
    </p>
  </div>
   
  <div
    class="notFound-container--column"
  >
    <h2
      class="notFound__information--header"
    >
      Eller sjekk ut en av disse sidene
    </h2>
     
    <p>
      
      (de fungerer, vi lover)
    
    </p>
     
    <div
      class="topics-grid"
    >
      <v-category-stub
        bgimage="/img/husholdninger"
        category="Befolkning"
        district="alle"
        id="husholdninger"
        topic="Husholdninger"
        txtcolor="rgb(245, 173, 165)"
      />
      <v-category-stub
        bgimage="/img/boligpriser"
        category="Boforhold"
        district="alle"
        id="boligpriser"
        topic="Boligpriser"
        txtcolor="rgb(199, 247, 201)"
      />
    </div>
  </div>
</div>
`;
